codestyle/styles.md

```diff
@@ -0,0 +1,672 @@
+* no base64 encoded images (it does a builder);
+* no `!important` directive;
+  * Except cases when you use it proactively. Proactive use of `!important`
+    means that it's used as a guarantee that some rule always will work, but not
+    as a fix of a specificity problem you’ve already encountered. For example:
+
+    ```CSS
+        .hidden {
+            display: none !important;
+        }
+    ```
+
+* no units for zero values: `margin: 0` is ok, but: `margin: 0px` is wrong;
+
+
+
+<a id="bem"></a>
+## BEM
+
+Describing the essence of BEM methodology is beyond the scope of
+this document, everything you need you can find on the official BEM site:
+[BEM methodology](https://en.bem.info/method/). Here are just described the
+formal rules of naming CSS classes.
+
+<a id="bem_naming"></a>
+### Naming conventions
+
+*Class name:*
+* minimum 3 symbols length;
+* always starts with a letter;
+* allowed characters: letters `(a-z)`, digits `(0-9)`, underscore `(_)`,
+  hyphen `(-)`;
+* hyphen `(-)` to separate words;
+* meaningful class names in English;
+* no word contractions:
+
+```CSS
+/* Bad choice — contractions */
+.lnk, .itm, .idx, .arr {}
+
+/* Good choice */
+.link, .item, .index, .arrow {}
+```
+
+* no abbreviations:
+
+```CSS
+/* Bad choice — abbreviations */
+.pdp, .fb, .pm {}
+
+/* Good choice */
+.product-page, .facebook, .panel-menu {}
+```
+
+* strive to use place-independent names. For example, instead of a class like
+`.site-nav`, choose something like `.primary-nav`; rather than
+`.footer-links`, favour a class like `.sub-links`.
+* no class names which describe block appearance, like `.blue` or `.border`.
+Instead, favour a class like `.highlight-color`.
+
+
+<a id="bem_block"></a>
+### Block
+
+Block is a common CSS class which has to meet common naming conventions,
+described above. The only exception: underscore `(_)` cannot be used in the
+block class name.
+
+```CSS
+.block {
+    /* declarations */
+}
+```
+
+for example:
+
+```CSS
+.button {
+    display: inline-block;
+    padding: 1em;
+    background: #587185;
+    color: #FFF;
+}
+```
+
+
+<a id="bem_element"></a>
+### Element
+
+The element class name has the following structure:
+
+1. the name of the block which owns the element;
+2. two (2) underscore signs `(__)` following the block name;
+3. the element name;
+
+```CSS
+.block__element {
+    /* declarations */
+}
+```
+
+for example:
+
+```CSS
+/* this is a block */
+.button {
+    ...
+}
+
+    /* this is an element which belongs to the block */
+    .button__icon {
+        display: inline-block;
+        margin-right: 5px;
+    }
+```
+
+<a id="bem_modifier"></a>
+### Modifier
+
+The modifier may refer to an element or block. The class name has the following
+structure:
+
+1. a block or an element class name;
+2. one (1) underscore sign `(_)`;
+3. the name of the modifier key;
+4. one (1) underscore sign `(_)`;
+5. the modifier value;
+
+```CSS
+.block_key_value {
```

Comment:
there are cases when it's better to use boolean modifiers (AKA modifiers without values). For example we have `tabs` block and it can be `tabs_dynamic` (with ability to add more tabs) and `tabs_rearrangeable` (with ability to drag-n-drop tabs). You can have `tabs_type_dynamic` and `tabs_type_rearrangeable`, but they can be dynamic and rearrangeable in the same time — so in this case we will have one modifiers with two values which is wrong. You can of course do something like `tabs_dynamic_true` and `tabs_rearrangeable_true`, but it's stupid and doesn't make any sense.